caffe2/ideep/operators/conv_transpose_op.cc

```diff
@@ -0,0 +1,241 @@
+#include "caffe2/operators/conv_transpose_op.h"
+#include "caffe2/ideep/operators/conv_transpose_unpool_base_op.h"
+#include "caffe2/ideep/operators/operator_fallback_ideep.h"
+#include <vector>
+
+namespace caffe2 {
+
+// TODO:it will be remove if bug of type zero pad be fixed in mkl-dnn.
+bool need_type_zero_pad(const mkldnn_memory_desc_t *pd) {
+  if (pd->ndims == 0)
+    return false;
+  int p1 = 1, p2 = 1;
+  for (int i = 0; i < pd->ndims; i++) {
+    p1 *= pd->dims[i];
+    p2 *= pd->layout_desc.blocking.padding_dims[i];
+  }
+  return (p1 != p2);
+}
+
+class IDEEPConvTransposeOp final : public IDEEPConvTransposeUnpoolBase {
+ public:
+  USE_IDEEP_DEF_ALIASES();
+  USE_IDEEP_CONV_TRANSPOSE_UNPOOL_BASE_FUNCTIONS();
+  // TODO:it will be remove if bug of type zero pad be fixed in mkl-dnn.
+  using FALLBACK_OP = IDEEPFallbackOp<ConvTransposeOp<float, CPUContext>>;
+
+  IDEEPConvTransposeOp(const OperatorDef& operator_def, Workspace* ws)
+      : IDEEPConvTransposeUnpoolBase(operator_def, ws),
+        training_mode_(
+            OperatorBase::GetSingleArgument<int>("training_mode", 0)),
+        fallback_(operator_def, ws) {
+    OPERATOR_NEEDS_FEATURE(
+        pad_l() == pad_r() && pad_t() == pad_b(),
+        "Uneven padding not supported.");
+  }
+  virtual ~IDEEPConvTransposeOp() {}
+
+  bool RunOnDeviceWithOrderNCHW() override {
+    const auto& X = Input(INPUT);
+    const auto& filter = Input(FILTER);
+    auto* Y = Output(OUTPUT);
+    CAFFE_ENFORCE_EQ(X.ndims(), 4);
+    CAFFE_ENFORCE_EQ(filter.ndims(), 4);
+    CAFFE_ENFORCE_EQ(filter.get_dim(2), kernel_h());
+    CAFFE_ENFORCE_EQ(filter.get_dim(3), kernel_w());
+
+    ideep::tensor::dims Y_dims;
+    bool pre_converted;
+    if (!(pre_converted = (filter.get_public_format() == ideep::format::iohw))) {
```

Comment:
Please don't combine assign operator in another part.
```suggestion
   const bool pre_converted = filter.get_public_format() == ideep::format::iohw;
   if (!pre_converted) {
```